app/src/main/java/org/mozilla/tv/firefox/ScreenControllerStateMachine.kt

```diff
@@ -0,0 +1,61 @@
+/* -*- Mode: Java; c-basic-offset: 4; tab-width: 4; indent-tabs-mode: nil; -*-
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.tv.firefox
+
+/**
+ * State machine for which Fragment is currently visible in the app.
+ */
+class ScreenControllerStateMachine {
```

Comment:
Strongly disagree.  If we ever want to substitute in a fake value for some test, having this be a class is highly preferred to making it a top-level object.  I don't think we should be using top-level objects anywhere outside of tests, tbh.

Reply:
> If we ever want to substitute in a fake value for some test,

@Baron-Severin Do you mean via mocking? You can mock `object` because, unlike Java's statics, they're a global singleton class instances. e.g.:
```kotlin
object UserAgentProvider {
    fun getUA() = ...
}

fun setUserAgent(userAgentProvider: UserAgentProvider) = ...

fun testUserAgent() {
    val uap = mock(UserAgentProvider::class.java)
    ...
```

The downsides to `object`, afaict, is that:
- If you want to initialize it with a value, it's preferable to use a constructor over static `init()` so you'll have to refactor it to a class
- It encourages static calling over passing it in as a dependency
- You can't extend/override it but you can implement an interface which can accomplish many of the same goals in a less error-prone way. With Kotlin's object delegation, this is only slightly more verbose than extending

I haven't formed a strong opinion on which approach is preferred.

Does this address your concerns?

>  I don't think we should be using top-level objects anywhere outside of tests, tbh.

Even with the above concern, I think there is some value in defining trivial top-level functions in `object` if they don't require the android framework (which can break mocking): e.g. a function re-used across the app to get the "representative character" from a URL, e.g. to sub in for home tile images.

Reply:
Objects can be mocked, but the fact that they can be accessed statically makes that difficult (and messy) in practice.  Client code generally isn't going to pass objects into their call sites (if they do, there isn't much of an advantage to using an object in the first place), so we usually won't have a good seam for subbing one out.  IMO, the service locator accomplishes the same goal but makes it easier to substitute values in and out when required.

Reply:
To summarize, it sounds like your concern is this point of mine:

> It encourages static calling over passing it in as a dependency

I think that's fair and I don't feel _strongly_ using one over the other. I'll keep it in mind and maybe default to class. Thanks for keeping me on my toes. :)

In this case, I think some functions in the class were defined statically: I think they should be defined as member functions to avoid this issue.